src/extension/wps/src/main/java/au/org/emii/gogoduck/worker/CSVConverter.java

```diff
@@ -0,0 +1,252 @@
+package au.org.emii.gogoduck.worker;
+
+import org.apache.commons.io.FilenameUtils;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.io.FileUtils;
+import org.geotools.data.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.commons.io.FileUtils.readFileToString;
+
+public class CSVConverter extends Converter {
+
+    private String ncksPath = null;
+    private String ncpdqPath = null;
+    private final String mimeType = "text/csv";
+    private final String extension = "csv";
+    private Integer lineNumber = 0;
+    private Integer attributesEnd = 0;
+    private List<String> lines = new ArrayList<String>();
+    private List<String> metadata = new ArrayList<String>();
+    private List<String> attributeMetadata = new ArrayList<String>();
+    private List<Map> attributes = new ArrayList<Map>();
+    private StringBuilder csvString = new StringBuilder();
+
+    private final String NEW_LINE = System.getProperty("line.separator");
+    private final String COMMENT = "# ";
+    private static final Logger logger = LoggerFactory.getLogger(GoGoDuck.class);
+
+    public CSVConverter() {}
+
+    @Override
+    public void init() {
+        this.ncksPath = GoGoDuckConfig.ncksPath;
+        this.ncpdqPath = GoGoDuckConfig.ncpdqPath;
+    }
+
+    @Override
+    public String getMimeType() { return mimeType; }
+
+    @Override
+    public String getExtension() { return extension; }
+
+    @Override
+    public Path convert(Path outputFile) throws GoGoDuckException {
+        List<String> ncpdqCommand = new ArrayList<>();
+        List<String> ncksCommand = new ArrayList<>();
+        String baseFilename = FilenameUtils.getBaseName(String.valueOf(outputFile));
+        try {
+            File tmpNcdf = File.createTempFile("tmp", ".nc");
+            // Unpack packed CSV values and store in temp csv
+            ncpdqCommand.add(ncpdqPath);
+            ncpdqCommand.add("-U");
+            ncpdqCommand.add("-O");
+            ncpdqCommand.add(outputFile.toString());
+            ncpdqCommand.add(tmpNcdf.getPath());
+            GoGoDuck.execute(ncpdqCommand);
+
+            // Output temp csv and store values
+            ncksCommand.add(ncksPath);
+            ncksCommand.add("--no_nm_prn");
+            ncksCommand.add("-a");
+            ncksCommand.add(tmpNcdf.getPath());
+
+            return executeNcks(ncksCommand, baseFilename, tmpNcdf.toPath());
```

Comment:
No race condition, you need to:
 * Store the result from `executeNcks`
 * Delete your temporary file
 * Return the result

It's tedious, but those practices will prevent you from forgetting things lying around. But look at my other comment about how to structure this function and keeping only high level logic in it.